### PR TITLE
Heat solver interface

### PR DESCRIPTION
**Pull request by dbochkov-flexcompute**
Since the old PR (https://github.com/flexcompute/tidy3d/pull/896) became very outdated, I've decided to open a new one.

Major changes/additions (directory structure is up for discussion, just whatever seemed to make sense for now):

### `components/heat_spec.py`: 
All medium classes can now have field `heat_spec` which can be either `SolidSpec` or `FuildSpec` for specifying thermal properties. To remind, there are three categories of properties:
1. optic (permittivity, conductivity, ...) - defined using standard `Medium`, `Lorentz`, etc classes
2. thermal (heat conductivity, capacity, ...) - added through field `heat_spec`.
3. response of optic properties to temperature (perturbations) - added when using `PerturbationMedium` and `PerturbationPoleResidue` classes.

And this is how different options interact with heat solver:
- Structures using medium instances with `heat_spec=None` or `heat_spec=FluidSpec()` will **not** be included in solution domain for the heat equation, and they will **not** be converted to custom mediums after applying temperature fields.
- Structures using non-perturbation medium instances with `heat_spec=SolidSpec(...)` will be included in solution domain for heat equation, but they will **not** be converted to custom mediums after applying temperature fields.
- Structures using perturbation medium instances (`PerturbationMedium` and `PerturbationPoleResidue`) with `heat_spec=SolidSpec(...)` will be included in solution domain for heat equation, and they will be converted to custom mediums after applying temperature fields. (the fullest material description)

### `components/scene.py`: 
Carved out geometric only information from `Simulation` into class `Scene` for easier transfer of geometric information between different solvers. Basically, it contains `structures`, `medium`, `center`, `size` + plotting functionality. 
- Point to discuss: plotting of custom media is dependent on simulation grid in `Simulation`. I think we should still have eps plotting in `Scene` class (along with thermal conductivity plotting, and perhaps other properties in future), so I adjusted the plotting of custom medium to use custom medium sample points + structure bounds instead of simulation grid. But now the question is whether `Simulation` in future should just directly borrow this plotting method or to have its own based on actual simulation grid?
- As discussed, for now we will not make any changes to `Simulation`, only added property `Simulation.scene` and class method `Simulation.from_scene(scene: Scene, **kwargs)` to transferring `Scene` between optic/heat simulation.
- Contrary, `HeatSimulation` does have field `scene: Scene` 

### `components/bc_placement.py`: 
This file defines classes for placement of boundary conditions on complex interfaces (`BCPlacementType=[StructureBoundary, StructureStructureInterface, MediumMediumInterface, etc]`). I realized this should be separated from heat solver components because it can be reused for other future solvers.

### `components/heat/`: 
The remaining of heat components are placed in this directory to keep everything tidy (yes, you see what I did here lol) and they include:
- `boundary.py` boundary conditions specifications: `TemperatureBC`, `HeatFluxBC`, `ConvectionBC` + `HeatBoundarySpec`
- `grid.py` heat grid specification: currently, `UniformHeatGrid` and `DistanceHeatGrid`. I can that this can go either way: we can introduce more classes to give the user more control or, contrary, unify everything into one spec so that the user don't have to think about what to select at all.
- `sim_data.py` defines `HeatSimulationData` which is currently simply `HeatSimulation` instance + temperature `SpatialDataArray`
- `simulation.py` defines `HeatSimulation`
- `source.py` heat source: currently only `UniformHeatSource`
- `viz.py` some parameters for plotting heat information

### Notebooks with demonstrations
There are two draft notebooks demonstrating:
- basic heat solver usage + comparison to analytical solutions https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/HeatSolverDraft.ipynb
- usage in the context of thermally tuning a ring resonator https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/ThermallyTunedRingResonatorDraft.ipynb
These are meant to provide more explanation on how every component is interacting with each other

### additional major things off the top of my head to discuss:
- Currently boundary conditions and heat source placement specifications are done through using structures and medium names. For example, `placement = td.StructureStructureInterface(structures=["box", "sphere"])`. This doesn't seem ideal. We could maybe just pass structures/mediums themselves. But that will increase size of `HeatSimulation` instances. Maybe an alternative is to just record hash of passed structures/mediums? But then it would be harder to point user which structure/medium is missing from a heat simulation.

### CODE REVIEW

File: tidy3d/components/heat/simulation.py
Comment (tylerflex) [inline tidy3d/components/heat/simulation.py:362]: this function is extremely long.. does it make sense to try to break it up into sub functions? It would be a lot of work and potentially more boilerplate / functions being used once but could make it easier to read and potentially more re-usable. I'll leave it up to you.

File: tidy3d/components/heat/source.py
Comment (tylerflex) [inline tidy3d/components/heat/source.py:21]: I might be mistaken but I think `Tuple[str]` means that it has to be a tuple of only **one** string, eg `("hello",)`.  I think what you might want is `Tuple[str, ...]` but double check me on this.

File: tidy3d/components/heat_spec.py
Comment (tylerflex) [inline tidy3d/components/heat_spec.py:40]: add "in units of" maybe

File: tidy3d/components/medium.py
Comment (tylerflex) [inline tidy3d/components/medium.py:188]: maybe we can add more description here. For example something like
```
If a heat_spec is added to a medium, it will be registered if the corresponding ``Structure`` is added to a ``HeatSimulation``.  <insert something about how to run heat sim and convert to optical sim>.  Otherwise, the ``heat_spec`` will not affect the running of an optical ``Simulation``.
```
Basically, just give a curious user looking at the medium API the opportunity to learn more about the heat solver.

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:39]: Maybe we can make this more clear, such as
```
Contains generic information about the geometry and medium properties common to all types of simulations.
```

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:123]: do we need this here? or should it just be in `Simulation`? I forget how this works or if we need it to be defined inside of a scene itself.

File: tidy3d/web/heat.py
Comment (tylerflex) [inline tidy3d/web/heat.py:1]: We'll clearly need to unify the webapis between all of the solvers because this is almost a direct copy of the `Simulation` webapi right? 

I'm working with MC to provide a simple interface where they provide a template for a `Submittable` task (simulation, mode solver, heat simulation) and we implement this template in our webapi and use a single webapi to manage it. We'll see how that looks but that would hopefully eliminate the need for this file.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Looks really good @dbochkov-flexcompute. Mostly minor comments and clarifications here.

File: tests/test_components/test_scene.py
Comment (dbochkov-flexcompute) [reply tests/test_components/test_scene.py]: good idea. In fact, I can just use `SCENE = SIM_FULL.scene` haha

File: tidy3d/components/bc_placement.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/bc_placement.py]: fixed

File: tidy3d/components/bc_placement.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/bc_placement.py]: fixed

File: tidy3d/components/heat/grid.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/grid.py:28]: it's only relevant for `Sphere` and `Cylinder`, as now explained in description

File: tidy3d/components/heat/sim_data.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/sim_data.py]: resolved

File: tidy3d/components/heat/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/simulation.py]: made this work

File: tidy3d/web/heat.py
Comment (dbochkov-flexcompute) [reply tidy3d/web/heat.py:1]: let's ignore web api stuff for now until stub is merged

**Comment by dbochkov-flexcompute** [comment]
Major revision taking into account all comment. Most significant additions are PolySlab support, Symmetry support, and proper monitor/simulation data classes. It seemed to accomplish that the most natural way would be to create abstract base classes for main components since their optic and heat specification will share many things, I put them in `base_sim/` for now. So, aiming at the following structure eventually for simulation classes (and similarly for other components):
```
      Scene
        |
  (composition)
        |
AbstractSimulation(ABC)------------------
        |                               |
  (inheritance)                   (inheritance)
        |                               |
Simulation(AbstractSimulation)    HeatSimulation(AbstractSimulation)
```

Demonstration notebooks have been updated too:
- https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/HeatSolverDraft.ipynb
- https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/heat-solver-draft/docs/source/notebooks/ThermallyTunedRingResonatorDraft.ipynb
These are meant to provide more explanation on how every component is interacting with each other


One concerns: we chose to just copy relevant code from `Simulation` to `Scene` for now, thinking that just removing duplicating code would be easier if everything works out. But I start noticing that `Simulation` functionalities start diverging from whatever I have put in `Scene`. This could happen for other components. So, perhaps, we shouldn't wait too long until we actually reimplement `Simulation` and other optic classes based on introduced in this PR `Scene` and higher level abstractions.

**Comment by momchil-flex** [comment]
Yes, that sounds important. Let's aim to have that done for 2.5 too. Maybe even do the refactor as soon as we start a 2.5 branch?

**Comment by momchil-flex** [comment]
I see, not so straightforward yeah, but in terms of backwards compatibility I think we'll need to do something like point 2. for now, where simulation is still initialized with center, size, medium, and then a scence is maybe initialized as a private attribute or is stored as a cached property or something.

**Comment by dbochkov-flexcompute** [comment]
there are currently 4 commits in this PR:
- "heat solver interface": last review state
- "heat solver web api": [new stuff] implementation of heat web api based on stub refactor
- "moving structures/medium back from ...": [new stuff]
-- removes `size` and `center` from `Scene` and makes them automatically calculated properties.
-- renames `UniformHeatGrid` and `DistanceHeatGrid` to `UniformUnstructureGrid` and `DistanceUnstructureGrid` since they could be reused in other solvers
-- makes `AbstractSimulation` be a derivative of `Box`, thus `domain` field is no longer needed. Simulation domains is set as usual through `size` and `center`.
-- add `structures` and `medium` to `AbstractSimulation`. + property `scene` + class method `from_scene`
- "temporarily limiting matplotlib ...": just a temporary commit to make it easier to deploy and test, can be removed at merge if no longer required

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:85]: just noticed, this should be Tidy3D (capital D)

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:95]: seems like a reasonable way to do this

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:169]: simulation -> scene (everywhere in file)

File: tidy3d/components/scene.py
Comment (tylerflex) [inline tidy3d/components/scene.py:257]: I think it's ok for now, but eventually we could probably put a lot of these static methods in viz.py or some `components/scene/viz.py` if we wanted to separate them out.

File: tidy3d/components/simulation.py
Comment (tylerflex) [inline tidy3d/components/simulation.py:3223]: is this implemented in the super class?

Comment (tylerflex) [reply tidy3d/components/simulation.py:3223]: same for `from_scene`? also add from_scene return type

File: tidy3d/components/validators.py
Comment (tylerflex) [inline tidy3d/components/validators.py:194]: quotes around field name

File: tidy3d/web/api/tidy3d_stub.py
Comment (tylerflex) [inline tidy3d/web/api/tidy3d_stub.py:32]: eventually we should add some more description here, if only to help other developers understand

File: tidy3d/web/api/webapi.py
Comment (tylerflex) [inline tidy3d/web/api/webapi.py:286]: I won't go through all of this logic, but we should test this pretty thoroughly because the webapi logic tends to break in weird ways sometimes.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Thanks @dbochkov-flexcompute. I think this looks really good. Mostly this is just minor stuff.

File: tidy3d/components/heat/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/simulation.py:93]: My understanding was that we should use discriminator whenever variable type is Union with more than 1 element, so that pydantic doesn't try to initialize provided argument with every element of that Union. Do you know if something changed during pydantic 2 transition?

File: tidy3d/components/heat/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/simulation.py:129]: support is planned after initial release

File: tidy3d/components/heat/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/simulation.py]: redefining it here to add example in docstring

File: tidy3d/components/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/simulation.py:3223]: actually, currently `Simulation` doesn't inherit from the new `AbstractSimulation` per our discussion to leave it on its own for a bit (similarly, `SimulationData`, monitors, etc). But I would like to do this transition in the next PR.

File: tidy3d/web/api/webapi.py
Comment (dbochkov-flexcompute) [reply tidy3d/web/api/webapi.py:292]: good idea, everything seemed to be working, except for download_log. but it was an issue on backend and now is fixed.

File: tidy3d/components/heat/simulation.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/simulation.py:399]: it will eventually go to `components/unstructured_sim`

File: tidy3d/components/bc_placement.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/bc_placement.py:1]: it will eventually go to `components/unstructured_sim`

File: tidy3d/components/heat/data/sim_data.py
Comment (dbochkov-flexcompute) [reply tidy3d/components/heat/data/sim_data.py:78]: yeah, I was hoping to just have a common implementation, but there were a bit too many specific things for each simulation type. Probably just leave like that for now and revisit after initial release

**Comment by dbochkov-flexcompute** [comment]
addressed comments + added a beta stage disclaimer in heat solver webapi:
> Tidy3D's heat solver is currently in the beta stage. All heat simulations are charged a flat fee of 0.025 FlexCredit.

File: tidy3d/components/heat/simulation.py
Comment (tylerflex) [reply tidy3d/components/heat/simulation.py:93]: I was under the impression that something changed with the discriminator recently, but I think I might be mistaken. let's leave it in.

File: tidy3d/components/heat/simulation.py
Comment (tylerflex) [reply tidy3d/components/heat/simulation.py:399]: a lot of plotting logic probably can be separated into other files, since it seems like a lot of them are just `staticmethod`s. But we can do this at a later time, I think.

File: tidy3d/web/api/webapi.py
Comment (tylerflex) [reply tidy3d/web/api/webapi.py:292]: have you tried job and batch of heat simulation? or even mixes of simulation and heat sim? it could be interesting to see how that works too. 

In my experience, sometimes the bugs were happening in `web.monitor()` in weird edge cases, for example 
* a task errors (if that happens we generally try the task again and see if it errors twice)
* the timing is somewhat different for whatever reason (task runs before the % done is ready)

There were some improvements earlier this year to fix some of these issues but if this PR changes any of that logic, they could appear again.  Will be good to run a bunch just to ensure they don't pop up again because they can be quite tricky to reproduce.

**Review by momchil-flex — COMMENTED** [review]
I'm pretty lost here so @dbochkov can you at least help me understand the current state and the plan for the general unification of things?

For example: HeatSimulation inherits from AbstractSimulation, but Simulation does not. Similarly HeatSimulationData inherits from AbstractSimulationData, but SimulationData does not. And while AbstractMonitorData has been introduced, both MonitorData and HeatMonitorData just inherit from Dataset.

Is there currently any duplicated code between Scene and Simulation?

I propose clarifying how everything is meant to work and putting the next few days of effort into actually merging this PR into 2.5 in a way that will avoid duplicated code and further code divergence. Then we can review/fix smaller details.

**Comment by dbochkov-flexcompute** [comment]
> I'm pretty lost here so @dbochkov can you at least help me understand the current state and the plan for the general unification of things?
> 
> For example: HeatSimulation inherits from AbstractSimulation, but Simulation does not. Similarly HeatSimulationData inherits from AbstractSimulationData, but SimulationData does not. And while AbstractMonitorData has been introduced, both MonitorData and HeatMonitorData just inherit from Dataset.
> 
> Is there currently any duplicated code between Scene and Simulation?
> 
> I propose clarifying how everything is meant to work and putting the next few days of effort into actually merging this PR into 2.5 in a way that will avoid duplicated code and further code divergence. Then we can review/fix smaller details.

At this stage `HeatMonitorData` should inherit from `AbstractMonitorData`, but `MonitorData` not. I'll fix that. 

So, the idea was to only have heat classes inheriting new `Abstract*` classes to test the new code organization while keeping optic classes untounched, in case something doesn't work out and we need to rethink it. I planned to do the removal of the duplicated code and switching optic classes to `Abstract*` classes in a separate PR. Are you suggesting just to go ahead and do that in this PR? I guess that would have the benefit of double-checking that nothing weird happens when proceeding with this unification.

**Comment by dbochkov-flexcompute** [comment]
Otherwise, we seemed to have converged on pretty much everything in this PR. Just need to test out `Batch` with failing heat tasks

**Comment by momchil-flex** [comment]
I think separate PR is fine. Just sooner rather than later. Generally as far as I'm concerned we could merge this soon (after you rebase again and answer all current comments, etc.) And just get it out there so that we can start getting some real user feedback and start fixing things. :)

**Comment by momchil-flex** [comment]
Is it currently possible for a user to have a look at the mesh that was used after a solver run?

**Comment by dbochkov-flexcompute** [comment]
> Is it currently possible for a user to have a look at the mesh that was used after a solver run? I guess they should at least be able to download the mesh file even if we don't provide any gmesh-wrapped functionality because of their GPL license.

yes, it's not ideal currently, but you can directly download mesh ("mesh.cgns") or solution+mesh ("volume_proc0.vtu") using `web.core.s3utils.download_file` to visualize in thirdparty tools like paraview or pyvista. 

This user experience can definitely be improved. Probably, once we have heat monitors that return data on the original unstructured grid instead of interpolated to an equivalent Cartesian one. One idea could be to introduce something like `MeshMonitor` that only return mesh in the selected region. And if a simulation contains only such monitors, then only meshing is performed and solver step is skipped. This will provide a way to inspect and iterate on mesh before actually solving heat equation. Or just provide meshing-only "solver" similar to mode solver.

File: requirements/basic.txt
Comment (dbochkov-flexcompute) [reply requirements/basic.txt]: removed

**Pull request merged**
